retail/interactive-tutorial/RetailProducts.Samples/SetInventorySample.cs

```diff
@@ -0,0 +1,126 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// [START retail_set_inventory]
+
+using Google.Cloud.Retail.V2;
+using Google.Protobuf.WellKnownTypes;
+using System;
+using System.Threading;
+
+/// <summary>
+/// The set inventory sample class.
+/// </summary>
+public class SetInventorySample
+{
+    /// <summary>
+    /// Get the retail product with inventory info.
+    /// </summary>
+    /// <param name="product">The actual product.</param>
+    /// <returns>Retail product with updated inventory info.</returns>
+    private static Product SetInventoryForProduct(Product product)
+    {
+        string[] placeIds = { "store1", "store2" };
+
+        product.FulfillmentInfo[0].Type = "pickup-in-store";
+        product.FulfillmentInfo[0].PlaceIds.Clear();
+        product.FulfillmentInfo[0].PlaceIds.AddRange(placeIds);
+        product.PriceInfo.Price = 15.0f;
+        product.PriceInfo.OriginalPrice = 20.0f;
+        product.PriceInfo.Cost = 8.0f;
+        product.PriceInfo.CurrencyCode = "USD";
+
+        return product;
+    }
+
+    /// <summary>
+    /// Get the set inventory request.
+    /// </summary>
+    /// <param name="product">The actual product.</param>
+    /// <returns>Set inventory request.</returns>
+    private static SetInventoryRequest GetSetInventoryRequest(Product product)
+    {
+        SetInventoryRequest setInventoryRequest = new SetInventoryRequest
+        {
+            Inventory = SetInventoryForProduct(product),
+            AllowMissing = true,
+            SetMask = new FieldMask
+            {
+                Paths = { "price_info", "availability", "fulfillment_info", "available_quantity" }
+            }
+        };
+
+        Console.WriteLine("Set inventory. request:");
+        Console.WriteLine($"Product name: {setInventoryRequest.Inventory.Name }");
+        Console.WriteLine($"Product fultillment info: {setInventoryRequest.Inventory.FulfillmentInfo}");
+        Console.WriteLine();
+
+        return setInventoryRequest;
+    }
+
+    /// <summary>
+    /// Call the Retail API to set product inventory
+    /// </summary>
+    /// <param name="product">The actual product.</param>
+    private static void SetProductInventory(Product product)
+    {
+        SetInventoryRequest setInventoryRequest = GetSetInventoryRequest(product);
+
+        ProductServiceClient client = ProductServiceClient.Create();
+        client.SetInventory(setInventoryRequest);
+
+        // This is a long running operation and its result is not immediately present with get operations,
+        // thus we simulate wait with sleep method.
+        Console.WriteLine("Set inventory. Wait 50 seconds:");
+        Console.WriteLine();
+
+        Thread.Sleep(50000);
```

Comment:
The library does all of this for you, you just need to do as follows:

```suggestion
 // Make the request
            Operation<SetInventoryResponse, SetInventoryMetadata> response = client.SetInventory(setInventoryRequest);

            // Poll until the returned long-running operation is complete
            Operation<SetInventoryResponse, SetInventoryMetadata> completedResponse = response.PollUntilCompleted();
            // Retrieve the operation result
            SetInventoryResponse result = completedResponse.Result;
            
            Console.WriteLine(....);       
```
See the **Example** on this [section of the reference docs](https://cloud.google.com/dotnet/docs/reference/Google.Cloud.Retail.V2/latest/Google.Cloud.Retail.V2.ProductServiceClient#Google_Cloud_Retail_V2_ProductServiceClient_SetInventory_Google_Cloud_Retail_V2_Product_Google_Protobuf_WellKnownTypes_FieldMask_Google_Api_Gax_Grpc_CallSettings_) for more info.

Reply:
Agree, this looks more generic, but after implementing this functionality, after every Api call "**response.PollUntilCompleted()**" I receive the next error message:

_Grpc.Core.RpcException: 'Status(StatusCode="PermissionDenied", Detail="The caller does not have permission", DebugException="Grpc.Core.Internal.CoreErrorDetailException: {"created":"@1647014879.849000000","description":"Error received from peer ipv4:142.250.186.170:443","file":"..\..\..\src\core\lib\surface\call.cc","file_line":1067,"grpc_message":"The caller does not have permission","grpc_status":7}")'_

However, the same method "**PollUntilCompleted()**" works fine for all Import samples. This error is received only for Add / Remove Fullfillment and SetInventory samples.

Maybe you have an idea of why is it happening to those samples?

Reply:
FYI: all other comments are fixed, and looks like averything is working fine except this operation.